README.md
Python MediaFire Client
=======================

Implements MediaFire Core API: http://www.mediafire.com/developers/core_api/

This is a work in progress. Expect breaking changes.

API Client library
------------------

API Client library provides a low-level interface to MediaFire API. It handles
requests, signatures and errors. Uses python-requests and requests_toolbelt for streaming uploads and does not require the whole file to
be loaded in memory.

Usage:

    from mediafire.api import MediaFireApi

    api = MediaFireApi()
    session_result = api.user_get_session_token(
        email='your.email@example.net',
        password='password',
        app_id='42511')

    api.set_session_token(session_result)

    response = client.user_get_info()
    print(response['user_info']['display_name'])

    # Or directly for methods that are not yet wrapped
    response = api.request("upload/add_web_upload", {
        "url": "http://www.mediafire.com/images/brand-assets/mf-logo.png",
        "filename": "mf-logo.png"})

    response = api.request("upload/get_web_uploads",
                        {"key": status['upload_key']})


API Client library supports operation w/o session_token. In this case all
operations that do require session_token will fail with Access denied error.

High-level library
------------------

High-level client library wraps API calls and presents simplified interface.

At the moment it supports:

 * Upload (instant, simple, and resumable, can use action tokens)
 * Download (direct download link)
 * Listing directories
 * Creating directories
 * Removing files and directories
 * Getting info about files and directories

MediaFire resources can be referenced by path or by quickkey/folderkey.

* path: mf:/Pictures/Sample.jpg or /Pictures/Sample.jpg
* folder_key: mf:6302u1a9p0a9x (folder_key is 13 chars long)
* quick_key: mf:46d3y4p8542kiyp (quick_key is 15 chars long)

examples/mediafire-cli.py is currently the only example of high level
client usage. Here's how to start:

    from mediafire.client import (MediaFireClient, File, Folder)

    client = MediaFireClient()
    client.login(email='your.email@example.net',
        password='password',
        app_id='42511')

    client.upload_file("flower.jpg", "mf:/Pictures/")
    client.download_file("mf:/Pictures/flower.jpg",
                         "flower-from-mediafire.jpg")

    contents = client.get_folder_contents_iter("mf:/Pictures")
    for item in contents:
        if type(item) is File:
            print("File: {}".format(item['filename']))
        elif type(item) is Folder:
            print("Folder: {}".format(item['foldername']))


CLI Interface
-------------

Work in progress. Can be used for basic tasks, such as directory listing,
uploads, downloads, getting resource information and removing files and folders
from the command line.


        usage: mediafire-cli [-h] [--debug] [--email EMAIL] [--password PASSWORD]
                         {ls,file-upload,file-download,folder-create,resource-delete,
                          file-update-metadata,folder-update-metadata,debug-get-resource}
                         ...

        Command-line interface to MediaFire Simple File Sharing and Storage

        optional arguments:
        -h, --help            show this help message and exit
        --debug               Enable debug output
        --email EMAIL
        --password PASSWORD

        Actions:
        {ls,file-upload,file-download,folder-create,resource-delete,file-update-metadata,
         folder-update-metadata,debug-get-resource}
        ls                  List directory contents
        file-upload         Upload files
        file-download       Download file
        folder-create       Create folder
        resource-delete     Delete resource
        file-update-metadata
                            Update file metadata
        folder-update-metadata
                            Update folder metadata
        debug-get-resource  Get resource info (debug)


Use --debug option to see API calls and internal debug information.

Requirements
------------

* python3
* requests
* requests\_toolbelt
* responses (for testing)

Installing
----------

I highly suggest installing it per-user or a virtualenv and avoid installing this library systemwide.

Clone it:

    $ git clone https://github.com/roman-yepishev/python-mediafire-sdk.git
    $ cd python-mediafire-sdk

For user installation, follow https://pip.pypa.io/en/latest/installing.html to get `get-pip.py`, then:

    $ PATH=~/.local/bin:$PATH
    $ python3 get-pip.py --user

    $ pip3 install -r requirements.txt -I
    $ python3 setup.py build install --user
    $ mediafire-cli --help

When in virtualenv, run this:

    $ pip3 install -r requirements.txt
    $ python3 setup.py build install

    # make sure that cli is working
    $ examples/mediafire-cli --help

Tests
-----

Basic test suite is located under tests/

Run it with:

    cd python-mediafire-sdk
    PYTHONPATH=. python3 -munittest


ABOUT AND LICENSE
Copyright (c) 2012-2013, Roman Yepishev. All right reserved. Website : rtg.in.au

This project was forked by MediaFire with explicit permission from Roman Yepishev on 10.24.2014

This project is made under BSD license. See LICENSE file for more information.

MediaFire® is a registered trademark of the MediaFire, LLC.


CHANGES.txt
2021-03-11 Release 0.7.0
 * API: Update API_VER to 1.5 - #14
 * improve performance for folders containing many entries by adding an option
   to pass folder/file path instead of quick_key/folder_key (which needs to be
   resolved first)
 * remove broken tests (image resize service returns 404)
 * obtian mediafire username from MEDIAFIRE_USERNAME env. variable
 * update comments

2016-11-15 Release 0.6.0
 * ! MediaFireUploader API breaking change: upload() no longer
   accepts hash_info argument - #30.
 * API:
   * Stringify numbers for request headers.

2016-01-31 Release 0.5.2
 * API:
   * Make sure to send JSON encoded to bytes - #22

2016-01-13 Release 0.5.1
 * MediaFireClient:
   * Fix creating folders recursively - #21
   * Fix URI parsing - #20

2016-01-09 Release 0.5.0
 * ! MediaFireUploader API breaking change: hash_ -> hash_info
 * Calculate unit_size and hash file/units in one go - #3
 * MediaFireClient:
   * Fix resource lookup for large folders - #17
   * Use explicit more_chunks flag from server - #9
   * Full path resolution support - #1
   * Force HTTPS downloads

2015-05-22 Release 0.4.1
 * Trap and retry upload exceptions - #16

2015-05-09 Release 0.4.0
 * API: Update API_VER to 1.3 - #14
 * Uploader: Control overwriting files with action_on_overwrite - #12
 * Media: Initial Conversion Server support - #5
 * Add integration tests - #15

2014-11-07 Release 0.3.0
 * Enable MediaFireApi and MediaFireUploader import from mediafire module
 * Use session property instead of function call, allow unsetting the token
 * Don't crash after FileDrop upload/simple, upload/resumable
 * Update README

2014-11-02 Release 0.2.0
 * Factor out uploader code
 * Python 2 support
 * Bug fixes and tests

2014-09-07 Release 0.1.0
 * Initial release.


examples/mediafire-cli.py
#!/usr/bin/env python3

"""Command-line interface to MediaFire Simple File Sharing and Storage"""

import argparse
import logging
import os
import sys

from pprint import pprint

from mediafire.client import (MediaFireClient,
                              Folder, ResourceNotFoundError)


LOG_FORMAT = '%(levelname)s %(asctime)s %(funcName)s: %(message)s'

logging.basicConfig(format=LOG_FORMAT, level=logging.WARNING)

APP_ID = '42511'


logging.getLogger("mediafire.client").setLevel(logging.INFO)
logging.getLogger(
    "requests.packages.urllib3.connectionpool").setLevel(logging.WARNING)


def do_ls(client, args):
    """List directory"""

    for item in client.get_folder_contents_iter(args.uri):
        # privacy flag
        if item['privacy'] == 'public':
            item['pf'] = '@'
        else:
            item['pf'] = '-'

        if isinstance(item, Folder):
            # type flag
            item['tf'] = 'd'
            item['key'] = item['folderkey']
            item['size'] = ''
        else:
            item['tf'] = '-'
            item['key'] = item['quickkey']
            item['name'] = item['filename']

        print("{tf}{pf} {key:>15} {size:>10} {created} {name}".format(**item))

    return True


def do_file_upload(client, args):
    """Upload files"""

    # Sanity check
    if len(args.paths) > 1:
        # destination must be a directory
        try:
            resource = client.get_resource_by_uri(args.dest_uri)
        except ResourceNotFoundError:
            resource = None

        if resource and not isinstance(resource, Folder):
            print("file-upload: "
                  "target '{}' is not a directory".format(args.dest_uri))
            return None

    with client.upload_session():
        for src_path in args.paths:
            print("Uploading {} to {}".format(src_path, args.dest_uri))
            result = client.upload_file(src_path, args.dest_uri)

            print("Uploaded {}, result={}".format(src_path, result))

    return True


def do_file_download(client, args):
    """Download file"""

    # Sanity check
    if not os.path.isdir(args.dest_path) and not args.dest_path.endswith('/'):
        print("file-download: "
              "target '{}' is not a directory".format(args.dest_path))
        if not os.path.exists(args.dest_path):
            print("\tHint: add trailing / to create one")
        return None

    for src_uri in args.uris:
        print("Downloading {} to {}".format(src_uri, args.dest_path))
        client.download_file(src_uri, args.dest_path)
        print("Downloaded {}".format(src_uri))

    return True


def do_file_show(client, args):
    """Output file contents to stdout"""
    for src_uri in args.uris:
        client.download_file(src_uri, sys.stdout.buffer)

    return True


def do_folder_create(client, args):
    """Create directory"""
    for folder_uri in args.uris:
        client.create_folder(folder_uri, recursive=True)
    return True


def do_resource_delete(client, args):
    """Remove resource"""
    for resource_uri in args.uris:
        client.delete_resource(resource_uri, purge=args.purge)
        print("Deleted {}".format(resource_uri))
    return True


def do_file_update_metadata(client, args):
    """Update file metadata"""
    client.update_file_metadata(args.uri, filename=args.filename,
                                description=args.description, mtime=args.mtime,
                                privacy=args.privacy)
    return True


def do_folder_update_metadata(client, args):
    """Update file metadata"""
    client.update_folder_metadata(args.uri, foldername=args.foldername,
                                  description=args.description,
                                  mtime=args.mtime, privacy=args.privacy,
                                  privacy_recursive=args.recursive)
    return True


def do_debug_get_resource(client, args):
    """Stat resource"""

    pprint(client.get_resource_by_uri(args.uri))
    return True


def main():  # pylint: disable=too-many-statements
    """Main entry point"""
    parser = argparse.ArgumentParser(prog='mediafire-cli',
                                     description=__doc__)

    parser.add_argument('--debug', dest='debug', action='store_true',
                        default=False, help='Enable debug output')
    parser.add_argument('--email', dest='email', required=False,
                        default=os.environ.get('MEDIAFIRE_EMAIL', None))
    parser.add_argument('--password', dest='password', required=False,
                        default=os.environ.get('MEDIAFIRE_PASSWORD', None))

    actions = parser.add_subparsers(title='Actions', dest='action')
    # http://bugs.python.org/issue9253#msg186387
    actions.required = True

    # ls
    subparser = actions.add_parser('ls',
                                   help=do_ls.__doc__)
    subparser.add_argument('uri', nargs='?',
                           help='MediaFire URI',
                           default='mf:///')

    # file-upload
    subparser = actions.add_parser('file-upload',
                                   help=do_file_upload.__doc__)
    subparser.add_argument('paths', nargs='+',
                           help='Path[s] to upload')
    subparser.add_argument('dest_uri', help='Destination MediaFire URI')

    # file-download
    subparser = actions.add_parser('file-download',
                                   help=do_file_download.__doc__)
    subparser.add_argument('uris', nargs='+',
                           help='MediaFire File URI[s] to download')
    subparser.add_argument('dest_path', help='Destination path')

    # file-show
    subparser = actions.add_parser('file-show',
                                   help=do_file_show.__doc__)
    subparser.add_argument('uris', nargs='+',
                           help='MediaFire File URI[s] to print out')

    # folder-create
    subparser = actions.add_parser('folder-create',
                                   help=do_folder_create.__doc__)
    subparser.add_argument('uris', nargs='+',
                           help='MediaFire folder path URI[s]')

    # resource-delete
    subparser = actions.add_parser('resource-delete',
                                   help=do_resource_delete.__doc__)
    subparser.add_argument('uris', nargs='+',
                           help='MediaFire resource URI[s]')
    subparser.add_argument('--purge', help="Purge, don't send to trash",
                           dest="purge", action="store_true", default=False)

    # file-update-metadata
    subparser = actions.add_parser('file-update-metadata',
                                   help=do_file_update_metadata.__doc__)
    subparser.add_argument('uri', help='MediaFire file URI')
    subparser.add_argument('--filename', help='Set file name',
                           default=None, dest='filename')
    subparser.add_argument('--privacy', help='Set file privacy',
                           choices=['public', 'private'],
                           default=None, dest='privacy')
    subparser.add_argument('--description',
                           help='Set file description',
                           dest='description', default=None)
    subparser.add_argument('--mtime', help="Set file modification time",
                           dest='mtime', default=None)

    # folder-update-metadata
    subparser = actions.add_parser('folder-update-metadata',
                                   help=do_folder_update_metadata.__doc__)
    subparser.add_argument('uri', help='MediaFire folder URI')
    subparser.add_argument('--foldername', help='Set folder name',
                           default=None, dest='foldername')
    subparser.add_argument('--privacy', help='Set folder privacy',
                           choices=['public', 'private'],
                           default=None, dest='privacy')
    subparser.add_argument('--recursive', help='Set privacy recursively',
                           action='store_true', default=None,
                           dest='recursive')
    subparser.add_argument('--description',
                           help='Set folder description',
                           dest='description', default=None)
    subparser.add_argument('--mtime', help='Set folder mtime',
                           default=None, dest='mtime')

    # debug-get-resource
    subparser = actions.add_parser('debug-get-resource',
                                   help=do_debug_get_resource.__doc__)
    subparser.add_argument('uri', help='MediaFire resource URI',
                           default='mediafire:/', nargs='?')

    args = parser.parse_args()

    if args.debug:
        logger = logging.getLogger()
        logger.setLevel(logging.DEBUG)

        logging.getLogger("mediafire.client").setLevel(logging.DEBUG)

    client = MediaFireClient()

    if args.email and args.password:
        client.login(args.email, args.password, app_id=APP_ID)

    router = {
        "file-upload": do_file_upload,
        "file-download": do_file_download,
        "file-show": do_file_show,
        "ls": do_ls,
        "folder-create": do_folder_create,
        "resource-delete": do_resource_delete,
        "file-update-metadata": do_file_update_metadata,
        "folder-update-metadata": do_folder_update_metadata,
        "debug-get-resource": do_debug_get_resource
    }

    if args.action in router:
        result = router[args.action](client, args)

        if not result:
            sys.exit(1)
    else:
        print('Unsupported action: {}'.format(args.action))
        sys.exit(1)

if __name__ == "__main__":
    main()
